### ISSUE TITLE
No offline whiteboards without login?

### ISSUE THREAD

**Issue by toobaz** [Y0ngg4n/fluffy_board#27]
The server is apparently very slow at the moment: the most annoying part is that until the app is able to load the list of whiteboards, it doesn't even display offline whiteboards... these should ideally be usable also without Internet connection (without even an account on any server).

**Comment by Y0ngg4n** [comment]
I will close this now. if you have any further questions please feel free to comment it here.

**Comment by toobaz** [comment]
> The problem why you need an account is the opportunity to upload your offline whiteboards and this can´t be done without an account.

Well, if I use "offline" whiteboards it means that having them online is not a priority ;-) While having them offline (even without login) is.

Anyway, I understand this is not a desired feature.

**Comment by Y0ngg4n** [comment]
@toobaz yes maybe i will implement it if i have free time

**Comment by ildar** [comment]
please reopen as a feature request.
It's really useless new without ANY working server instance because a new user cannot login even once.

**Comment by Y0ngg4n** [comment]
@ildar currently i don´t have the time for this project. And it need a big refactoring.

**Comment by MoralCode** [comment]
Currently this seems to be the only quality whiteboard app on fDroid. 

I would love to contribute some time to making the app work entirely locally for people like me who want to use it for quick sketches and simple notes on the go.

Can you elaborate on the large refactor you said this would require? Would love to learn more about what might be needed to add this functionality

**Comment by Y0ngg4n** [comment]
@MoralCode ok so the problem currently is that i am saving the data in json files and i have to figure out a more performant way to save the data. I think the best way would be to migrate to [Isar](https://github.com/isar/isar) but that needs a complete rewrite of the current schemas.
Also the scaling of items needs a full rewrite.
And there is something wrong with the ci pipeline causing f-droid not using the latest version.
Also there are still  some performance problems on lower end devices that need to be addressed.

I would really appreciate any help, but currently i don´t have much time for this project.

**Comment by MoralCode** [comment]
A lot of these sound like they are valid issues but may not be directly related to the offline whiteboards issue so i broke them out into new issues to help keep track of them more easily. More details on each of those would be nice to have, but i'll focus on the offline one in this issue.

How is data currently being saved? are you just storing JSON files to the phones storage? or is the process of saving/loading the JSON data dependent on the server in some way?

I guess ultimately my question is, could I theoretically just delete all the server-dependent stuff, like the login screen, and make the app mostly work in an offline-only way, or otherwise delay the process of asking for a login until the user takes an action that cannot be done without connecting to their account (like clicking a button to "upload" a sketch)?

**Comment by Y0ngg4n** [comment]
@MoralCode thank you :)

Ok so the current problem why you can´t use it without an account is because there are some checks missing regarding the offline usage.
Yes i am just saving the json files to the phones storage with the localstorage plugin. But this plugin has its own issues on other devices than phones because it just drop the json files directly in the documents folder. 
I had choosen the json files back then, because they make it easy to import a whiteboard. 
So i think for exports the json format is great, just for the saving of the whiteboard offline it does not seem to fit the use case.

So the json file is always saved by <uuid>.json in the documents folder. but there is also another json file (just currently can´t remember the location) that is just listing all the uuids of the whiteboards, because without this file i would not be able to find the offline whiteboards as the localstorage plugin does not provide a search functionality.

So for the offline work without account i think it will be necessary to add those many checks if you are registered or not or if you are just offline currently. The rewrite to isar is currently not necessary to work without an account but needs to be done anyways.

**Comment by MoralCode** [comment]
So all thats needed is to wrap the online-dependent lines of code in a check/prompt to ask the user if they want to use online functionality or just skip that feature?

**Comment by MoralCode** [comment]
Note to self: https://github.com/Y0ngg4n/fluffy_board/blob/master/lib/dashboard/dashboard.dart is where the check and redirect to the login page on app start is performed, this is probably one place that should be modified

**Comment by Y0ngg4n** [comment]
yes exactly. There is already a isOffline bool i have added. So the dashboard needs those checks and also the file manager, because there needs to be a limited functionality too.

**Comment by MoralCode** [comment]
Does this sound like what you had in mind?

treat offline mode as the default and have there be points in the app (such as on first startup, or whenever a users account is needed) where the app will 
1) check if the user has previously logged in (i.e. if there are saved credentials or something) and use those
2) if not, prompt for credentials with the option to skip.
3) if user provides credentials, continue using the online stuff  if not, store the users offline preference in a (maybe this is what isOffline is) to avoid over-notifying them in future

**Comment by Y0ngg4n** [comment]
@MoralCode yes that sounds great! :+1: 

The uploading and sharing of boards i think will be limited

**Comment by topialla** [comment]
@MoralCode did you make any progress on this already? Perhaps I could offer some time as well.

**Comment by Y0ngg4n** [comment]
@topialla i currently dont have time to maintain this project

**Comment by emmggi** [comment]
Is there any reason why it wasn't implemented with offline use from the start?

Also what kind of functionality and features with regards to drawing/sketching did this app have? Basic shapes, alignment, stroke color and width, text etc? I'm trying to adapt it to another app.

**Comment by Y0ngg4n** [comment]
offline use is implemented but it needs a registered account for sync functionality. Shapes. scribbles, highlighter, colors, text, pdf import, images and other stuff.

**Comment by emmggi** [comment]
Yes but I don't want sync. Needing a registered account even if you continue to use it offline only is a bit pointless.

If I understood correctly I need to use the docker image or set up a server to be able to register?

**Comment by Y0ngg4n** [comment]
Yes you need to install both server and client
